core/test/scenario/aggregate.go

```diff
@@ -0,0 +1,139 @@
+package scenario
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/get-eventually/go-eventually/core/aggregate"
+	"github.com/get-eventually/go-eventually/core/event"
+	"github.com/get-eventually/go-eventually/core/version"
+)
+
+// AggregateRootInit is the entrypoint of the Aggregate Root scenario API.
+//
+// An Aggregate Root scenario can either set the current evaluation context
+// by using Given(), or test a "clean-slate" scenario by using When() directly.
+type AggregateRootInit[I aggregate.ID, T aggregate.Root[I]] struct {
+	typ aggregate.Type[I, T]
+}
+
+// AggregateRoot is a scenario type to test the result of methods called
+// on an Aggregate Root and their effects.
+//
+// These methods are meant to produce side-effects in the Aggregate Root state, and thus
+// in the overall system, enforcing the domain invariants represented by the
+// Aggregate Root itself.
+func AggregateRoot[I aggregate.ID, T aggregate.Root[I]](typ aggregate.Type[I, T]) AggregateRootInit[I, T] {
+	return AggregateRootInit[I, T]{
+		typ: typ,
+	}
+}
+
+func (sc AggregateRootInit[I, T]) Given(events ...event.Persisted) AggregateRootGiven[I, T] {
+	return AggregateRootGiven[I, T]{
+		typ:   sc.typ,
+		given: events,
+	}
+}
+
+func (sc AggregateRootInit[I, T]) When(fn func() (T, error)) AggregateRootWhen[I, T] {
```

Comment:
I thought about it but I wanted to keep the API event-based (i.e. `Given` being a list of Domain Events applied over the Aggregate), rather than function based.

Especially if `Then()` asserts on Domain Events.

Reply:
The `aggregate.Type` already provides the factory for creating empty `T` instances, but the reason why `AggregateRootInit.When()` is different from `AggregateRootGiven.When()` is because, with `Given`, there are preconditions to be set.

Do you think the state-based API (as you suggested) makes more sense over the event-based one?

Reply:
I think it makes sense to still use event-based approach. My point was mostly about UX.
Maybe there shouldn't even be `AggregateRootInit.When`? I.e. always force building Aggregate with Events and then allow to do some additional checks on the Aggregate.

Reply:
>Maybe there shouldn't even be AggregateRootInit.When? I.e. always force building Aggregate with Events and then allow to do some additional checks on the Aggregate.

How can we test the case of, for example, `user.Create()`?

Reply:
Would you mind post a concrete example? You can use the "User" example in the repository 😄

Reply:
You won't like it :D
If some limitations are acceptable and we can assume that even if some Command Handler affects multiple Aggregate types, the scenario covers only one of them (kind of like having a single mock + multiple stubs), perhaps something like this is possible with Go generics:
```go
scenario.
  AggregateRoot(user.Type). // defines in-memory aggregate repository type
  CommandHandler[user.CreateCommand, user.CreateCommandHandler]().
  // .Given(events...) // creates an empty Aggregate, applies events, stores aggregates
  When(command.Envelope[user.CreateCommand]{
    Message: user.CreateCommand{
	FirstName: "John",
	LastName:  "Doe",
	BirthDate: now,
	Email:     "john@doe.com",
    },
    Metadata: nil,
  }).
  ExpectEvents(events...). // it's not Given-When-Then, but it's more readable IMO
  ExpectAggregate(func (aggregate *user.User) { // aggregate can be detected from in-memory tracer
    assert.Equal(t, "John", aggregate.FirstName)
    // checks whatever is important in the Aggregate
  }).
  AssertOn(t, func (repo aggregate.Repository[user.User]) user.CreateCommandHandler {
    // build Command Handler using provided repository 
  })
```
It's not concrete, sorry, but I hope the idea is clear.